src/headers.jl

```diff
@@ -0,0 +1,246 @@
+# See https://httpwg.github.io/specs/rfc7230.html#rule.token.separators
+const validHeaderBytes = IntSet(Int(c) for c in
+   raw"!#$%&'*+-.0123456789ABCDEFGHIJKLMNOPQRSTUWVXYZ^_`abcdefghijklmnopqrstuvwxyz|~"
+)
+
+validHeaderChar(c::Char) = isascii(c) && UInt8(c) in validHeaderBytes
+
+function canonicalHeaderKey(s::String)
+    # fast path for valid header keys
+    upper = true
+    for c in s
+        if !validHeaderChar(c)
+            return s
+        end
+        if upper && 'a' <= c <= 'z'
+            return _chk(s)
+        end
+        if !upper && 'A' <= c <= 'Z'
+            return _chk(s)
+        end
+
+        upper = c == '-'
+    end
+
+    return s
+end
+
+const toUpper = 'A' - 'a'
+
+function _chk(s::String)
+    # confirm its fixable before changing anything
+    for c in s
+        if !validHeaderChar(c)
+            return s
+        end
+    end
+
+    a = copy(Array{UInt8}(s))
+    upper = true
+    for (i, b) in enumerate(a)
+        if upper && UInt8('a') <= b <= UInt8('z')
+            b += toUpper
+        elseif !upper && UInt8('A') <= b <= UInt8('Z')
+            b -= toUpper
+        end
+        @inbounds a[i] = b
+        upper = b == UInt8('-')
+    end
+
+    return String(a)
+end
+
+"""
+Internal type used to enable fast merge, copy, and filter of Headers
+"""
+immutable AlreadyCanonical
+    d::Dict{String,String}
+end
+
+"""
+    Headers() <: Associative{String,String}()
+
+A type to represent headers. Keys containing invalid characters are returned as is. Valid keys are canonicalized, uppercasing the first character and characters following a '-' and lowercasing other characters. So, for valid keys, Headers is case-insensitive.
+"""
+immutable Headers{K<:String,V<:String} <: Associative{K,V}
```

Comment:
SubString would be a little weird because canonicalization would replace arbitrary keys with new strings only to SubString them. I don't have any other string types in mind, but Id be happy to loosen the signature if anyone brought one.